cpp/src/strings/replace/replace.cu

```diff
@@ -146,34 +122,38 @@ namespace {
  *
  * This will perform a replace_slice operation on each string.
  */
-template <two_pass Pass = two_pass::SIZE_ONLY>
 struct replace_slice_fn {
   column_device_view const d_strings;
   string_view const d_repl;
-  size_type start, stop;
-  const int32_t* d_offsets{};
+  size_type const start;
+  size_type const stop;
+  int32_t* d_offsets{};
   char* d_chars{};
 
-  __device__ size_type operator()(size_type idx)
+  __device__ void operator()(size_type idx)
   {
-    if (d_strings.is_null(idx)) return 0;  // null string
-    string_view d_str = d_strings.element<string_view>(idx);
-    auto length       = d_str.length();
-    char* out_ptr     = nullptr;
-    if (Pass == two_pass::EXECUTE_OP) out_ptr = d_chars + d_offsets[idx];
-    const char* in_ptr = d_str.data();
-    size_type bytes    = d_str.size_bytes();
-    size_type begin    = ((start < 0) || (start > length) ? length : start);
-    size_type end      = ((stop < 0) || (stop > length) ? length : stop);
-    begin              = d_str.byte_offset(begin);
-    end                = d_str.byte_offset(end);
-    bytes += d_repl.size_bytes() - (end - begin);
-    if (Pass == two_pass::EXECUTE_OP) {
-      out_ptr = copy_and_increment(out_ptr, in_ptr, begin);
-      out_ptr = copy_string(out_ptr, d_repl);
-      out_ptr = copy_and_increment(out_ptr, in_ptr + end, d_str.size_bytes() - end);
+    if (d_strings.is_null(idx)) {
+      if (!d_chars) d_offsets[idx] = 0;
+      return;
+    }
+    auto const d_str   = d_strings.element<string_view>(idx);
+    auto const length  = d_str.length();
+    char const* in_ptr = d_str.data();
+    auto const begin   = d_str.byte_offset(((start < 0) || (start > length) ? length : start));
+    auto const end     = d_str.byte_offset(((stop < 0) || (stop > length) ? length : stop));
+    auto const bytes   = d_str.size_bytes() + d_repl.size_bytes() - (end - begin);
+
+    if (d_chars) {
+      char* out_ptr = d_chars + d_offsets[idx];
+
+      out_ptr = copy_and_increment(out_ptr, in_ptr, begin);  // copy beginning
+      out_ptr = copy_string(out_ptr, d_repl);                // insert replacement
+      out_ptr = copy_and_increment(out_ptr,                  // copy end
+                                   in_ptr + end,
+                                   d_str.size_bytes() - end);
+    } else {
+      d_offsets[idx] = bytes;
```

Comment:
```suggestion
    if (d_chars) {
       auto const d_str   = d_strings.element<string_view>(idx);
       auto const length  = d_str.length();
       char const* in_ptr = d_str.data();
       auto const begin   = d_str.byte_offset(((start < 0) || (start > length) ? length : start));
       auto const end     = d_str.byte_offset(((stop < 0) || (stop > length) ? length : stop));
    
      char* out_ptr = d_chars + d_offsets[idx];

      out_ptr = copy_and_increment(out_ptr, in_ptr, begin);  // copy beginning
      out_ptr = copy_string(out_ptr, d_repl);                // insert replacement
      out_ptr = copy_and_increment(out_ptr,                  // copy end
                                   in_ptr + end,
                                   d_str.size_bytes() - end);
    } else {
      d_offsets[idx] = d_str.size_bytes() + d_repl.size_bytes() - (end - begin);
```

Feel free to ignore if you think this would de-clarify the code at all.

Reply:
That is fine. You overdid though since `begin` and `end` are both needed in the `if` and `else` clauses. I will move the `bytes` calculation into just the `else`.

Reply:
Ah, right.  Should have looked more carefully.